stackslib/src/net/httpcore.rs

```diff
@@ -0,0 +1,1625 @@
+        _key: &StacksPublicKey,
+        _preamble: &StacksHttpPreamble,
+        _bytes: &[u8],
+    ) -> Result<(), NetError> {
+        // not defined for HTTP messages, but maybe we could add a signature header at some point
+        // in the future if needed.
+        Ok(())
+    }
+
+    /// Write out a message to `fd`.
+    ///
+    /// NOTE: If we're sending a StacksHttpMessage::Request(..), then the next preamble and payload
+    /// received _must be_ a StacksHttpMessage::Response(..) in response to the request.
+    /// If it is not, then that decode will fail.
+    fn write_message<W: Write>(
+        &mut self,
+        fd: &mut W,
+        message: &StacksHttpMessage,
+    ) -> Result<(), NetError> {
+        match *message {
+            StacksHttpMessage::Request(ref req) => {
+                // client cannot send more than one request in parallel
+                if self.request_handler_index.is_some() {
+                    test_debug!("Have pending request already");
+                    return Err(NetError::InProgress);
+                }
+
+                // find the response handler we'll use
+                let (decoded_path, _) = decode_request_path(&req.preamble().path_and_query_str)?;
+                let handler_index = self
+                    .find_response_handler(&req.preamble().verb, &decoded_path)
+                    .ok_or(NetError::SendError(format!(
+                        "No response handler found for `{} {}`",
+                        &req.preamble().verb,
+                        &decoded_path
+                    )))?;
+
+                req.send(fd)?;
+
+                // remember this so we'll know how to decode the response.
+                // The next preamble and message we'll read _must be_ a response!
+                self.request_handler_index = Some(handler_index);
+                Ok(())
+            }
+            StacksHttpMessage::Response(ref resp) => resp.send(fd),
+            StacksHttpMessage::Error(_, ref resp) => resp.send(fd),
+        }
+    }
+}
+
+impl PeerNetwork {
+    /// Send a (non-blocking) HTTP request to a remote peer.
+    /// Returns the event ID on success.
+    pub fn connect_or_send_http_request(
+        &mut self,
+        data_url: UrlString,
+        addr: SocketAddr,
+        request: StacksHttpRequest,
+    ) -> Result<usize, NetError> {
+        PeerNetwork::with_network_state(self, |ref mut network, ref mut network_state| {
+            PeerNetwork::with_http(network, |ref mut network, ref mut http| {
+                match http.connect_http(
+                    network_state,
+                    network,
+                    data_url.clone(),
+                    addr.clone(),
+                    Some(request.clone()),
+                ) {
+                    Ok(event_id) => Ok(event_id),
+                    Err(NetError::AlreadyConnected(event_id, _)) => {
+                        match http.get_conversation_and_socket(event_id) {
```

Comment:
Because this is a match on a singular match, you could conver this to an if let (Some(ref mut convo), Some(ref mut socket)) = BLAH if you want to. I think clippy might actually complain about this, but regardless...is just a syntactical preference.

Reply:
Fixed